README.md
# MNXB01-PROJECT
This is the project in the course MNXB01 2021 HT

All codes are based on Aurora. 

Load and run ROOT inside the directory src/ 

All of the files will be linked automatically.

We use the following commands to load ROOT on Aurora:

```
module load GCC/5.4.0-2.26  
module load OpenMPI/1.10.3  
module load ROOT/6.10.02-Python-2.7.12
```

Team members: Carl-Fredrik Lidgren, Daniel Falkowski, Isak Ellmer, Qian Liu 




https://www.overleaf.com/4442485513qrkytxnthsqr
the link for an online latex editor to the project report







ChangeLog
03:11:2021
Added features to Gregorian and WeatherDataVec classes, and compactified the code in corona.C -- DF
Fixed some bugs -- CF

02:11:2021
Added "global_temperature_regression.C", which tries to fit a linearly increasing sin function to given data.
Modified some of the utility classes to add some functionality. --CF

Added isempty to WeatherDataVec -- DF

Added more features to WeatherDataVec in particular avgtemp, mintemp, maxtemp --DF

Added a new class called Gregorian, this permits immediate addition of dates just initialize with Gregorian gDate {year,month,day};
Then you can freely add dates with gDate1 + gDate2. --DF

Added some more ways of accessing WeatherDataVec -- There are now two indexing operators [],(): 
--For WeatherDataVec Wdata we could have Wdata[5] or Wdata(5) both of which will return the 6th entry in data.
Furthermore Wdata(2,19) gives all entries from 2-19. Also, added Wdata.get_between() which takes two string arguments for the start and end date;
--and yields all dates in the intermediate range. --DF

Improve the classes Analyse_Monthly and Analyse_Corona, adding more parameters. --QL


01:11:2021
Added a ROOT script called "avg_periodtemp.C" which contains a function "temperature_over_two_periods" taking 2 pairs of ints and a WeatherDataVec
as input. Using each pair of ints, it constructs a histogram containing the average temperature for each day of the period defined by the two ints,
then plots the resulting two histograms on top of each other. It also computes the average difference in temperature between the two time periods.
Also updated README.md with some improvements in formatting and a fixed typo.  --CF

Added some more methods to WeatherDataVec and WeatherDataLine, these are mostly for convenience such as a means of extracting data by regex--
and overloading the index operator [], in a similar vein I added the long overdue constructor method for WeatherDataVec immediately from filename. --DF

28:10:2021
Encapsulated my functions from "monthly.C" into two classes:
class Analyse_Monthly    in    "monthly.h" & "monthly.C"
class Analyse_Corona    in    "corona.h" & "corona.C"
and improve rootlogon.C
--QL

24:10:2021
upload the 1st version root function to achieve the goals
the datapath in rootlogon.C may be modified since it is the absolute address to my system --QL


22:10:2021 

I added some headers today. Firstly, the header "csvregex.h" parses the raw data via regex. Secondly, the headers "WeatherDataLine.h" and "WeatherDataVec.h" -- 
provide nice wrappers for the data we are dealing with, in order to initialize the structures correctly just do something like the following:

//filename is the path to one of our raw ".csv" files
WeatherDataVec Wdata {datavec(filename)};

This provides immediate access to the data as a vector of WeatherDataLine, which encode the info of one line of the processed csv. To access the info within WeatherDataLine --
Just use one of the corresponding functions such as get_day().

IMPORTANT NOTE: it may very well be prudent to split up the header files into both .h and .cpp files, as they are currently quite short I deemed it unecessary;
and I think that we can keep the file for the WeatherDataLine class and for the miscellaneous functions for the csv parsing as is since I don't imagine they will grow in length.
However, we should consider whether or not to make a separate .cpp file for WeatherDataVec; this is probably the right thing to do depending on how we want to continue coding. -DF



src/monthly.c
#include <iostream>
#include <fstream>
#include <algorithm>
#include <tuple>
#include "csvregex.h"
#include "WeatherDataLine.h"
#include "WeatherDataVec.h"
using namespace std;

#include <TF1.h>
#include <TH1.h>
#include <TStyle.h>
#include <TGraph.h>
#include <TCanvas.h>
#include <TFile.h>
#include <TTree.h>
#include <TClonesArray.h>
#include <TVirtualFitter.h>
#include <TLatex.h>
#include "RootClass.C"

void temPerMonth_tree(Int_t month) // ttree built
{
	TFile* file = new TFile("monthdata.root", "RECREATE");
	TTree* tree = new TTree("tree", "Output tree");
	Year* byyear = new Year();
	tree->Branch("byyear", &byyear);
	TClonesArray* dataArray = new TClonesArray("TempData", 1000);
	tree->Branch("data", "TClonesArray", &dataArray);
	Int_t nT = 0;
	
	string filename {"../datasets/smhi-opendata_1_53430_20210926_101122_Lund.csv"};
	WeatherDataVec Wdata {datavec(filename)};
	WeatherDataVec datamonth = Wdata.get_by_month(month);
	vector<int> list_years = datamonth.list_years();
	for(Int_t n = 0; n < (Int_t) list_years.size() ; n++){
		byyear->year = list_years[n];

		WeatherDataVec _data = datamonth.get_by_year(byyear->year);
		
		byyear->nData = (Int_t) _data.data.size();

		Double_t Temp[byyear->nData];
		for(Int_t i = 0; i < byyear->nData; i++){
		Temp[i] = _data.data[i].get_temp();
		TempData* TD = new((*dataArray)[i]) TempData();
		nT++;
		TD->temp = Temp[i];
		}
	tree->Fill();
	}
	file->Write();
	file->Close();
}

void temPerMonth()//plot the histo of the average temperature of the month
{

	TFile* file = TFile::Open("monthdata.root");
	TTree* tree = (TTree*)file->Get("tree");
	Year* byyear = 0;
	TClonesArray* dataArray = 0;
	tree->SetBranchAddress("byyear", &byyear);
	tree->SetBranchAddress("data", &dataArray);
	

	TH1D* h1 = new TH1D("h1", "Monthly temperatures distribution; Celsius; Counts", 100, -10, 30);
	
	const Int_t nYears = tree->GetEntries();
	for (Int_t n = 0; n < nYears; n++) {
		tree->GetEntry(n);
		const Int_t ndata = dataArray->GetEntries();	
		Double_t TempSum = 0.0;
		for (Int_t i = 0; i < ndata; i++)
		{
			TempData* TD = (TempData*)dataArray->At(i);
			TempSum += TD->temp;
		}
		Double_t TempAver = TempSum/ndata;
		
		h1->Fill(TempAver);
	}
	h1->Fit("gaus");
	h1->Draw();

	TH1D *h2 = new TH1D("h2", "Monthly temperatures distribution; Temperature ^{o}C; Year Count", 100, -10, 30);
	(TVirtualFitter::GetFitter())->GetConfidenceIntervals(h2);
	h2->SetStats(kFALSE);
	h2->SetFillColor(41);
	h2->Draw("e3");
	h1->Draw("same");
}

void MonthExtreme()//plot the extreme temperatures & average temperature of the month
{
	TFile* file = TFile::Open("monthdata.root");
	TTree* tree = (TTree*)file->Get("tree");
	Year* byyear = 0;
	TClonesArray* dataArray = 0;
	tree->SetBranchAddress("byyear", &byyear);
	tree->SetBranchAddress("data", &dataArray);
	
	const Int_t nYears = tree->GetEntries();
	Double_t x[nYears], yH[nYears], yC[nYears], yA[nYears];
	
	for (Int_t n = 0; n < nYears; n++) {
		tree->GetEntry(n);
		x[n] = byyear->year;
		const Int_t ndata = dataArray->GetEntries();	
		Double_t HTemp = 0.0;
		Double_t CTemp = 0.0;
		Double_t TempSum = 0.0;
		for (Int_t i = 0; i < ndata; i++)
		{
			TempData* TD = (TempData*)dataArray->At(i);
			TempSum += TD->temp;
			if (i==0)
			{
				HTemp = TD->temp;
				CTemp = TD->temp;
			}
			else
			{
				HTemp = max(TD->temp, HTemp);
				CTemp = min(TD->temp, CTemp);
			}
		}
		Double_t TempAver = TempSum/ndata;
		yA[n] = TempAver;
		yH[n] = HTemp;
		yC[n] = CTemp;
	}
	
	TCanvas *c1 = new TCanvas("c1","Month Extreme Temperature",200,10,1000,600);
	TGraph* g1 = new TGraph(nYears, x, yH);
	TGraph* g2 = new TGraph(nYears, x, yC);
	TGraph* g3 = new TGraph(nYears, x, yA);
	TMultiGraph *mg = new TMultiGraph();
	mg->SetTitle("Month Extreme Temperature; Year; Temperature ^{o}C");
	g1->SetFillColor(46);
	g2->SetFillColor(38);
	g3->SetLineWidth(3);
	g3->SetLineColor(12);
	g3->Fit("pol1");
	mg->Add(g1);
	mg->Add(g2);
	mg->Draw("AB");
	mg->GetXaxis()->SetTitleSize(0.03);
	mg->GetXaxis()->SetLabelSize(0.03);
	mg->GetYaxis()->SetTitleSize(0.03);
	mg->GetYaxis()->SetLabelSize(0.03);
	g3->Draw("C same");
}

tuple<int, int, int> dateCalculation(int year, int month, int date, int period)//determine the day+period from y.m.d 
{
	date = date + period;
	int count =1;
	while(count !=0){
		count =0;
		if (date>31 && ( month==1 || month==3 || month==5 || month==7 || month==8 || month==10 || month==12 ) ){
			date = date - 31;
			month++;
			count++;
		}
		else if (date>30 &&  ( month== 4 || month==6 || month==9 || month==11 ) ){
			date = date - 30;
			month++;
			count++;
		}
		else if (date>28 && month ==2 && (year%4 !=0)){
			date = date - 28;
			month++;
			count++;
		}
		else if (date>29 && month ==2 && (year%4 ==0)){
			date = date - 29;
			month++;
			count++;
		}

		if (month > 12){
			month = 1;
			year++;
		}
	}

	return make_tuple(year, month, date);

}

Double_t meanOfPeriod (WeatherDataVec data, int frYear, int frMonth, int frDate, int period)// determine the mean temp from fY,fM,fD with step length period
{
// from date included, to date not included.
	Double_t sum=0.0;
	Int_t count=0;
	while(period!=0){
		WeatherDataVec _data = data.get_by_year(frYear);
		WeatherDataVec __data = _data.get_by_month(frMonth);
		WeatherDataVec ___data = __data.get_by_day(frDate);
		for(unsigned i=0; i<___data.data.size(); i++){
			sum = sum + ___data.data[i].get_temp();
			count++;
		}
		tie(frYear, frMonth, frDate) = dateCalculation(frYear, frMonth, frDate, 1);	
		period--;	
	}
	cout << sum/count << endl;
	return sum/count;
}


void Corona()//discover the relation between the temperature and corona
{
	ifstream file("../datasets/lund_corona_data");
	Int_t veck1[58];
	Int_t antal[58];
	Double_t veck2[58];

	Int_t i1 = 0;
	Double_t i2 = 0.0;
	Int_t year, week, kod, a, b, c, d;
	string lund;
	string _temp;
	while(file >> _temp)
	{
		if (_temp == "nya_fall_vecka")
		{
			break;
		}
	}

	while(file >> year >> week >> kod >> lund >> lund >> a >> b >> c >> d)
	{	
		if( (year == 2020 && week >= 9) || (year == 2021 && week <=13) )
		{
			veck1[i1] = i1;
			veck2[i1] = i2;
			antal[i1] = d;
			i1++;
			i2 = i2 + 1.0;
		}
	}


	string filename {"../datasets/smhi-opendata_1_53430_20210926_101122_Lund.csv"};
	WeatherDataVec Wdata {datavec(filename)};

	Double_t temp[58];
	////////this is for test, saving time
	//for (int i = 0; i < 58; i++)
	//temp[i] = 14.4;
	//////////////////////////
	Int_t frYear = 2020;
	Int_t frMonth = 2;
	Int_t frDate = 24;
	Int_t toYear = 2021;
	Int_t toMonth = 4;
	Int_t toDate = 4;
	Int_t _i = 0;
	while (!(frYear >= toYear && frMonth >= toMonth && frDate >= toDate)){
		cout<< frYear << ":" << frMonth<< " :" << frDate << endl;
		temp[_i] = meanOfPeriod(Wdata, frYear, frMonth, frDate, 7);
		tie(frYear, frMonth, frDate) = dateCalculation(frYear, frMonth, frDate, 7);
		_i++;
	}

	TCanvas *c1 = new TCanvas("c1","Temperature & Corona",200,10,1200,600);
	TGraph* g1 = new TGraph(58, veck1, antal);
	TGraph* g2 = new TGraph(58, veck2, temp);
	TPad *pad1 = new TPad("pad1","",0,0,1,1);
	TPad *pad2 = new TPad("pad2","",0,0,1,1);
	pad2->SetFillStyle(4000); // transparent
	pad2->SetFrameFillStyle(0);
	g1->SetFillColor(46);
	g1->SetTitle("Temperature & Corona");
	g1->GetYaxis()->SetLabelColor(46);
	g1->GetYaxis()->SetLabelSize(0.03); // default 0.04
	g1->GetYaxis()->SetTitle("Counts");
	g1->GetYaxis()->SetTitleColor(46);
	g1->GetYaxis()->SetTitleSize(0.03); // default 0.04
	g1->GetXaxis()->SetRangeUser(0, 62);
	g1->GetXaxis()->SetTitle("Weeks");
	g1->GetXaxis()->SetTitleSize(0.03); // default 0.04
	g1->GetXaxis()->SetLabelSize(0.03); // default 0.04
	g1->GetXaxis()->ChangeLabel(1,-1,-1,-1,-1,-1,"2020/9");
	g1->GetXaxis()->ChangeLabel(2,-1,-1,-1,-1,-1,"19");
	g1->GetXaxis()->ChangeLabel(3,-1,-1,-1,-1,-1,"29");
	g1->GetXaxis()->ChangeLabel(4,-1,-1,-1,-1,-1,"39");
	g1->GetXaxis()->ChangeLabel(5,-1,-1,-1,-1,-1,"49");
	g1->GetXaxis()->ChangeLabel(6,-1,-1,-1,-1,-1,"2021/6");
	g1->GetXaxis()->ChangeLabel(7,-1,-1,-1,-1,-1,"16");	
	g2->SetLineColor(9);
	g2->SetLineWidth(3);
	g2->GetXaxis()->SetTickLength(0);// transparent
	g2->GetXaxis()->SetLabelOffset(999);// transparent
	g2->GetXaxis()->SetLabelSize(0);// transparent
	g2->GetYaxis()->SetLabelColor(9);
	g2->GetYaxis()->SetLabelSize(0.03); // default 0.04
	g2->GetYaxis()->SetTitle(" ^{o}C");
	g2->GetYaxis()->SetTitleColor(9);
	g2->GetYaxis()->SetTitleSize(0.03); // default 0.04
	/*g2->GetYaxis()->SetTickLength(0);
	g2->GetYaxis()->SetLabelOffset(999);
	g2->GetYaxis()->SetLabelSize(0);*/
	pad1->Draw();
	pad1->cd();
	g1->Draw("AB, Y+");
	pad2->Draw();
	pad2->cd();
	g2->Draw();

}

